src/ucp/core/ucp_am.c

```diff
@@ -1784,6 +1784,22 @@ const ucp_request_send_proto_t ucp_am_reply_proto = {
 void ucp_am_proto_request_zcopy_reset(ucp_request_t *request)
 {
     ucs_assert(request->send.msg_proto.am.header.reg_desc != NULL);
+
+    /* If user header is not guaranteed to be valid, 
+     * use mpool buffer for storing the user header.
+     */
+    if ((request->send.msg_proto.am.flags & UCP_AM_SEND_FLAG_COPY_HEADER) &&
+        !(request->flags & UCP_REQUEST_FLAG_USER_HEADER_COPIED) &&
+        (request->send.msg_proto.am.header.length != 0)) {
+        request->send.msg_proto.am.header.ptr = ucs_mpool_set_get_inline(
+                &request->send.ep->worker->am_mps,
+                request->send.msg_proto.am.header.length);
+        request->flags |= UCP_REQUEST_FLAG_USER_HEADER_COPIED;
+    }
+
+    memcpy(request->send.msg_proto.am.header.ptr,
+           request->send.msg_proto.am.header.reg_desc + 1,
+           request->send.msg_proto.am.header.length);
```

Comment:
in case the if in line 1791 was false, what would be the value of request->send.msg_proto.am.header.ptr here?